### PR TITLE
Auto refresh view

### PR DESCRIPTION
**Pull request by mpscholten**
https://twitter.com/digitallyinduce/status/1308678484735283200

Open tasks:
- [x] Remove hard coded table names
- [x] Only send view when it actually changed
- [x] Improve handling of multiple events (currently one event leads to one rerender, instead we could also wait 100ms after the first event and then rerender which avoids rerenders for all events that would happen inbetween)
- [x] Documentation

### CODE REVIEW

**Comment by mpscholten** [comment]
@HugoPeters1024  maybe you could also take a look over this :) ?

**Comment by HugoPeters1024** [comment]
@mpscholten I didn't know up until this moment that postgres could do this (very cool indeed!) and find it hard to reason about the implementation. However, I will certain try to make an MVP using this functionality and let know you if it worked.

**Comment by mpscholten** [comment]
Thanks for the fast response. Cool, this is also a good for the documentation I guess :)

**Review by Laobiz — CHANGES_REQUESTED** [review]
Great work 👍 At the moment there is no delete trigger for the tables. It would make sense to also auto reload when a deletion happened.

File: Guide/auto-refresh.markdown [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+## How It Works
+
+It's good to have a general understanding of how IHP Auto Refresh works.
+
+Auto Refresh first have to be activated for an action by calling `autoRefresh`. Once activated the framework will automatically track all tables your action is using e.g. in `SELECT * FROM ...` queries. Once the action sends a response IHP will start watching for any kind of `INSERT` or `UPDATE` statement to all the tables used by your action.
+
+When the page is rendered a small javascript function will connect back to the IHP server using a WebSocket connection.
+
+Whenenver a `INSERT` or `UPDATE` happens to the tables used by your action IHP will rerun your action on the serverside. When the generated html looks different to the html generated on the initial page load it will send the new html to the browser using the WebSocket connection. The javascript listening on the WebSocket will use the new html to update the current page. It uses morphdom to only touch the parts of your current DOM that have actually changed.
```
Comment (HugoPeters1024) [inline Guide/auto-refresh.markdown:24]: I saw delete has now also been added :)

File: lib/IHP/static/ihp-auto-refresh.js [hunk 2]
```diff
@@ -10,17 +11,18 @@ function autoRefreshView() {
         return;
     }
 
+    var socketHost = "ws://" + window.location.hostname + ":" + document.location.port;
```
[the remark this answers is not in the document]
Comment (mpscholten) [reply lib/IHP/static/ihp-auto-refresh.js]: Yes that actually happend when we deployed it the first time to IHP cloud :D That was then fixed in a later commit via https://github.com/digitallyinduced/ihp/pull/419/commits/4cec09d8d59b612ebe592d54a1b28ead40692fff

File: Guide/auto-refresh.markdown (hunk 1 above)
Comment (mpscholten) [reply Guide/auto-refresh.markdown]: Good catch :) Fixed 👍

File: Guide/auto-refresh.markdown [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+### FrontController
+
+First you need to add it to your `Web/FrontController.hs`:
+
+```haskell
+module Web.FrontController where
+
+import IHP.AutoRefresh -- <------ ADD THIS IMPORT
+
+instance InitControllerContext WebApplication where
```
[the remark this answers is not in the document]
Comment (mpscholten) [reply Guide/auto-refresh.markdown:39]: Good point, just changed it 👍

**Pull request merged**